saleor/checkout/fetch.py

```diff
@@ -290,95 +284,85 @@ def fetch_checkout_info(
 ) -> CheckoutInfo:
     """Fetch checkout as CheckoutInfo object."""
+    channel = checkout.channel
+    shipping_address = checkout.shipping_address
+    if shipping_channel_listings is None:
+        shipping_channel_listings = channel.shipping_method_listings.all()
+
+    delivery_method_info = get_delivery_method_info(None, shipping_address)
     checkout_info = CheckoutInfo(
         checkout=checkout,
         user=checkout.user,
-        channel=checkout.channel,
+        channel=channel,
         billing_address=checkout.billing_address,
-        shipping_address=checkout.shipping_address,
-        delivery_method_info=DeliveryMethodBase(),
-        shipping_method_channel_listings=None,
-        valid_shipping_methods=[],
+        shipping_address=shipping_address,
+        delivery_method_info=delivery_method_info,
+        all_shipping_methods=[],
         valid_pick_up_points=[],
     )
-
-    if fetch_delivery_methods:
-        populate_checkout_info_shippings(checkout_info, lines, discounts, manager)
-        valid_pick_up_points = get_valid_collection_points_for_checkout_info(
-            checkout.shipping_address, lines, checkout_info
-        )
-        checkout_info.valid_pick_up_points = valid_pick_up_points
+    update_delivery_method_lists_for_checkout_info(
+        checkout_info,
+        checkout.shipping_method,
+        checkout.collection_point,
+        shipping_address,
+        lines,
+        discounts,
+        manager,
+        shipping_channel_listings,
+    )
 
     return checkout_info
 
 
-def populate_checkout_info_shippings(
+def update_checkout_info_delivery_method_info(
     checkout_info: CheckoutInfo,
-    lines: Iterable[CheckoutLineInfo],
-    discounts: Iterable["DiscountInfo"],
-    manager: "PluginsManager",
+    shipping_method: Optional[ShippingMethod],
+    collection_point: Optional[Warehouse],
+    shipping_channel_listings: Iterable[ShippingMethodChannelListing],
 ):
+    """Update delivery_method_attribute for CheckoutInfo.
+
+    The attribute is lazy-evaluated avoid external API calls unless accessed.
+    """
     from .utils import get_external_shipping_id
 
+    delivery_method: Optional[Union[ShippingMethodData, Warehouse, Callable]] = None
     checkout = checkout_info.checkout
-    channel = checkout_info.channel
-    shipping_address = checkout_info.shipping_address
-    shipping_method = checkout.shipping_method
-
-    external_shipping_methods = (
-        get_valid_external_shipping_method_list_for_checkout_info(
-            checkout_info, shipping_address, lines, discounts, manager
+    if shipping_method:
+        # Find listing for the currently selected shipping method
+        for listing in shipping_channel_listings:
+            if listing.shipping_method_id == shipping_method.id:
+                shipping_channel_listing = listing
+                break
+
+        delivery_method = convert_to_shipping_method_data(
+            shipping_method, shipping_channel_listing
```

Comment:
In theory yes, practically there's `drop_invalid_shipping_methods_relations_for_given_channels` that's supposed to clear `checkout.shipping_method` whenever the listing is dropped .

The issue should be gone once those 2 tables are denormalised with `CheckoutShippingMethod` so I think there's no need for a temporary workaround here.

Reply:
The exception is going to happen somewhere else, as we've decided to make the `amount` field mandatory in the schema. The only way to avoid exceptions we would have to make amount nullable once again.